### PR TITLE
fix: stop calling end in createDupeReqFilterMiddleware.ts, so that dapps get correct response to 'seenRequests'

### PR DESCRIPTION
**Pull request by danjm**
## **Description**

This PR fixes the following problem:

1. Dapp sends a request to MetaMask shortly after the service worker starts up, and _before_ the inpage provider receives the `METAMASK_EXTENSION_CONNECT_CAN_RETRY` message from the content-script (which happens after the metamask controller initializes and notifies all connections of a `chainChanged` event) 
2. The request hits the middleware created by `createDupeReqFilterMiddleware.ts` and its id is added to `seenRequestIds`
3. Before the necessary controller responds to the request, the inpage provider _now_ receives the `METAMASK_EXTENSION_CONNECT_CAN_RETRY` message from the content-script
4. The provider now retries the request. This happens without the dapp doing anything; this is part of our MV3 retry logic to ensure requests don't get lost under service worker stoppage / start-up conditions. (The provider does this via the imported `createStreamMiddleware` https://github.com/MetaMask/json-rpc-middleware-stream/blob/main/src/createStreamMiddleware.ts#L128-L130)
5. The new retry of the request hits the middleware created by `createDupeReqFilterMiddleware`. (The original request still has not been responded to.) The new retry of the request has the same id as the original, so the middleware hits the `} else if (!seenRequestIds.add(req.id)) {` condition and calls `end()`
6. The original request, which was being awaited by the dapp, now is resolved but without a meaningful response (as the middleware just did an `end()` call without the request being handled in any way) 

This problem was discovered by some e2e tests which became very flaky under MV3. Tests that involved the Simple Snap Keyring dapp would often send a `wallet_requestSnaps` request, and then not get the necessary response, due to the issue described above.

The solution to the problem presented in this PR is just to not call `end` when seeing a duplicate request, because the original request should be responded to eventually.

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/24672?quickstart=1)

## **Related issues**

Part of the resolution to https://github.com/MetaMask/metamask-extension/issues/21496

## **Manual testing steps**

The bug is hard to repro manually. You could, perhaps, locally modify the `createStreamMiddleware` code to call `sendToStream` a second time for the same request after a short (e.g. 10 millisecond) timeout (perhaps here: https://github.com/MetaMask/json-rpc-middleware-stream/blob/main/src/createStreamMiddleware.ts#L50-L60). On develop that should result in the original request receiving an empty response but on this branch the original request should receive a successful response.

This branch will also contribute to MV3 e2e tests passing, but that will require some other PRs as well.

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: app/scripts/lib/createDupeReqFilterMiddleware.ts [hunk 1]
```diff
@@ -52,14 +52,14 @@ export default function createDupeReqFilterMiddleware(): JsonRpcMiddleware<
...
+  return function filterDuplicateRequestMiddleware(req, _res, next) {
     if (req.id === undefined) {
       // JSON-RPC notifications have no ids; our only recourse is to let them through.
-      return next();
-    } else if (!seenRequestIds.add(req.id)) {
+      next();
+    } else if (seenRequestIds.add(req.id)) {
+      next();
+    } else {
       log.info(`RPC request with id "${req.id}" already seen.`);
```
Comment (rekmarks) [inline app/scripts/lib/createDupeReqFilterMiddleware.ts:62]:
```suggestion
      log.info(`Ignoring JSON-RPC request with already seen id "${req.id}".`);
```

File: app/scripts/lib/createDupeReqFilterMiddleware.ts [hunk 2]
```diff
@@ -52,14 +52,14 @@ export default function createDupeReqFilterMiddleware(): JsonRpcMiddleware<
   void
 > {
   const seenRequestIds = makeExpirySet();
-  return function filterDuplicateRequestMiddleware(req, _res, next, end) {
+  return function filterDuplicateRequestMiddleware(req, _res, next) {
```
Comment (rekmarks) [inline app/scripts/lib/createDupeReqFilterMiddleware.ts:55]: More explicit more better, maybe.
```suggestion
  return function filterDuplicateRequestMiddleware(req, _res, next, _end) {
```

**Review by rekmarks — COMMENTED** [review]
I believe I understand the problem, but I'm not sure if we can use this solution. While it will create the desired behavior, it will result in a perpetually hung promise for each duplicate request (see `@metamask/json-rpc-engine` [here](https://github.com/MetaMask/core/blob/3165e87096300a66e06bd949dde736f90ca24fbc/packages/json-rpc-engine/src/JsonRpcEngine.ts#L524) and [here](https://github.com/MetaMask/core/blob/3165e87096300a66e06bd949dde736f90ca24fbc/packages/json-rpc-engine/src/JsonRpcEngine.ts#L554)).

I think we need to set a bogus `res.id` and `end()` the request (I think ideally with an error, but it doesn't really matter) in order to hit [one of these cases in `createStreamMiddleware()`](https://github.com/MetaMask/core/blob/3165e87096300a66e06bd949dde736f90ca24fbc/packages/json-rpc-middleware-stream/src/createStreamMiddleware.ts#L107-L116) on the dapp side. In this way, the response to the duplicate request will be ignored on the dapp side, while the response to the original request will be handled as normal.

_Edit:_ I also discovered this bug in `json-rpc-middleware-stream`: https://github.com/MetaMask/core/issues/4308

File: app/scripts/lib/createDupeReqFilterMiddleware.ts [hunk 3]
```diff
@@ -52,14 +52,14 @@ export default function createDupeReqFilterMiddleware(): JsonRpcMiddleware<
...
       // JSON-RPC notifications have no ids; our only recourse is to let them through.
-      return next();
-    } else if (!seenRequestIds.add(req.id)) {
+      next();
+    } else if (seenRequestIds.add(req.id)) {
+      next();
+    } else {
       log.info(`RPC request with id "${req.id}" already seen.`);
-      return end();
     }
```
Comment (rekmarks) [inline app/scripts/lib/createDupeReqFilterMiddleware.ts:63]: Per [my review](https://github.com/MetaMask/metamask-extension/pull/24672#pullrequestreview-2072557424), here's what I think will work:
```suggestion
    res.id = null;
    return end(new Error(`Received request with duplicate id "${req.id}".`));
    }
```

**Comment by Gudahtt** [comment]
Perhaps it would be simpler to move this "ignore duplicate requests" step to _before_ we pass the request into the JSON-RPC engine. Then we'd have no hung promises, and we wouldn't need to mangle the request or artificially end

**Comment by danjm** [comment]
> Perhaps it would be simpler to move this "ignore duplicate requests" step to before we pass the request into the JSON-RPC engine.

Do you mean we would do that from within the provider? I probably don't understand something here... the `retryStuckRequests` call happens in the stream middleware after the request has been passed to the JSON-RPC engine. If we move this "ignore duplicate requests" logic to before that, then couldn't duplicate requests make it to the extension background?

Or do you mean that we will handle these duplicate requests within the metamask background / controller, after the message has been received from the dapp, but before it is streamed into the engine?

**Comment by legobeat** [comment]
> I believe I understand the problem, but I'm not sure if we can use this solution. While it will create the desired behavior, it will result in a perpetually hung promise for each duplicate request (see `@metamask/json-rpc-engine` [here](https://github.com/MetaMask/core/blob/3165e87096300a66e06bd949dde736f90ca24fbc/packages/json-rpc-engine/src/JsonRpcEngine.ts#L524) and [here](https://github.com/MetaMask/core/blob/3165e87096300a66e06bd949dde736f90ca24fbc/packages/json-rpc-engine/src/JsonRpcEngine.ts#L554)).


In case it makes a difference, those links are to latest version of `@metamask/json-rpc-engine` `8.0.2`, while metamask-extension is still using much older version `json-rpc-engine` `6.1.0`.

- Corresponding source file for `json-rpc-engine@6.1.0` [here](https://github.com/MetaMask/json-rpc-engine/blob/v6.1.0/src/JsonRpcEngine.ts).
- Pending upgrade to `@metamask/json-rpc-engine` pending in https://github.com/MetaMask/metamask-extension/pull/22875
  - Blocked by upgrading from `eth-rpc-errors` to `@metamask/json-rpc-errors`

**Comment by rekmarks** [comment]
> Or do you mean that we will handle these duplicate requests within the metamask background / controller, after the message has been received from the dapp, but before it is streamed into the engine?

@danjm IMO it would have to be this. I agree with @Gudahtt that this would be in many ways a more elegant solution. It could essentially be a middleware in the form of a `through` / `Transform` stream that we tack on to the stream pipeline before the `providerStream` on the inbound side only, [here](https://github.com/MetaMask/metamask-extension/blob/38199bb1dabf676200eec3ed8c40d3ee2ace51f3/app/scripts/metamask-controller.js#L4929).

**Comment by danjm** [comment]
I just did a force push, but I left the original 3 commits unchanged, I was just handling the rebase on latest develop

**Comment by danjm** [comment]
> this would be in many ways a more elegant solution. It could essentially be a middleware in the form of a through / Transform stream that we tack on to the stream pipeline before the providerStream on the inbound side only,

@Gudahtt @rekmarks I have done this in the latest commit https://github.com/MetaMask/metamask-extension/pull/24672/commits/84e947f965932f516b1a6ae88181f11e20542ac6

**Comment by danjm** [comment]
@metamaskbot update-policies

**Review by rekmarks — DISMISSED** [review]
All my comments are nits. Nice tests, LGTM!

File: app/scripts/lib/createDupeReqFilterThroughStream.ts [hunk 4]
```diff
@@ -42,24 +42,23 @@ const makeExpirySet = () => {
...
  * Ignores JSON-RPC notifications, i.e. requests with an `undefined` id.
  *
- * @returns The middleware function.
+ * @returns The stream object.
  */
-export default function createDupeReqFilterMiddleware(): JsonRpcMiddleware<
-  unknown,
-  void
-> {
+export default function createDupeReqFilterThroughStream() {
```
[the remark this answers is not in the document]
Comment (danjm) [reply app/scripts/lib/createDupeReqFilterThroughStream.ts]: Done in 42ef9da805

File: app/scripts/lib/createDupeReqFilterThroughStream.test.ts [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+      { id: 1, method: 'foo' },
+    ];
+
+    jest.advanceTimersByTime(THREE_MINUTES);
+
+    requests2.forEach((request) => testStream.write(request));
+    expect(output).toEqual(expectedOutputAfterExpiryTime);
+  });
+
+  it('does not expire single id after two minutes', () => {
```
[the remark this answers is not in the document]
Comment (danjm) [reply app/scripts/lib/createDupeReqFilterThroughStream.test.ts]: Updated in 96f050e8b5

File: app/scripts/lib/createDupeReqFilterStream.ts [hunk 6]
```diff
@@ -42,24 +42,23 @@ const makeExpirySet = () => {
...
+export default function createDupeReqFilterStream() {
   const seenRequestIds = makeExpirySet();
-  return function filterDuplicateRequestMiddleware(req, _res, next, end) {
-    if (req.id === undefined) {
-      // JSON-RPC notifications have no ids; our only recourse is to let them through.
-      return next();
-    } else if (!seenRequestIds.add(req.id)) {
-      log.info(`RPC request with id "${req.id}" already seen.`);
-      return end();
+  return createThroughStream((chunk, _, cb) => {
```
Comment (legobeat) [inline app/scripts/lib/createDupeReqFilterStream.ts:52]: This is just a wrapper for `Transform` in `readable-stream` already. Can we just use that directly?

https://github.com/rvagg/through2/blob/ebe3d0b736115ffca785336301119979df70a708/through2.js#L70

**Comment by danjm** [comment]
@metamaskbot update-policies

File: app/scripts/lib/createDupeReqFilterStream.ts (hunk 6 above)
Comment (danjm) [reply app/scripts/lib/createDupeReqFilterStream.ts]: Done in https://github.com/MetaMask/metamask-extension/pull/24672/commits/a8a25a0cc29058e114445f03bece395c53b52e00

File: app/scripts/lib/createDupeReqFilterStream.ts [hunk 7]
```diff
@@ -1,5 +1,6 @@
-import { JsonRpcMiddleware } from 'json-rpc-engine';
+import { Transform } from 'stream';
 import log from 'loglevel';
+import { JsonRpcRequest } from '@metamask/utils';
```
Comment (rekmarks) [inline app/scripts/lib/createDupeReqFilterStream.ts:3]:
```suggestion
import type { JsonRpcRequest } from '@metamask/utils';
```

File: app/scripts/metamask-controller.js [hunk 8]
```diff
@@ -4997,23 +4997,31 @@ export default class MetamaskController extends EventEmitter {
       tabId,
     });
 
+    const dupeReqFilterThroughStream = createDupeReqFilterStream();
```
Comment (rekmarks) [inline app/scripts/metamask-controller.js:5000]:
```suggestion
    const dupeReqFilterStream = createDupeReqFilterStream();
```

File: app/scripts/lib/createDupeReqFilterStream.ts [hunk 9]
```diff
@@ -42,24 +43,26 @@ const makeExpirySet = () => {
 };
 
 /**
- * Returns a middleware that filters out requests whose ids we've already seen.
+ * Returns a "through" stream that filters out requests whose ids we've already seen.
```
Comment (rekmarks) [inline app/scripts/lib/createDupeReqFilterStream.ts:46]:
```suggestion
 * Returns a transform stream that filters out requests whose ids we've already seen.
```

File: app/scripts/lib/createDupeReqFilterStream.test.ts [hunk 10]
```diff
@@ -0,0 +1,315 @@
+import { Transform } from 'stream';
+import { JsonRpcRequest } from '@metamask/utils';
```
Comment (rekmarks) [inline app/scripts/lib/createDupeReqFilterStream.test.ts:2]:
```suggestion
import type { JsonRpcRequest } from '@metamask/utils';
```

File: app/scripts/lib/createDupeReqFilterStream.test.ts [hunk 11]
```diff
@@ -0,0 +1,315 @@
...
+import createDupeReqFilterStream, {
+  THREE_MINUTES,
+} from './createDupeReqFilterStream';
+
+function createTestStream(output: JsonRpcRequest[] = []) {
+  const throughStream = createDupeReqFilterStream();
```
Comment (rekmarks) [inline app/scripts/lib/createDupeReqFilterStream.test.ts:8]:
```suggestion
  const transformStream = createDupeReqFilterStream();
```

**Review by rekmarks — COMMENTED** [review]
We could touch up some variable names since we stopped using `through2`.

File: app/scripts/lib/createDupeReqFilterStream.test.ts [hunk 12]
```diff
@@ -0,0 +1,315 @@
+import { Transform } from 'stream';
```
Comment (legobeat) [inline app/scripts/lib/createDupeReqFilterStream.test.ts:1]: @danjm Is there any reason why you prefer runtime-specific `stream` over suggested and deterministic `readable-stream` here?

Comment (danjm) [reply app/scripts/lib/createDupeReqFilterStream.test.ts]: oops, no

fixed in d20349a332

File: app/scripts/lib/createDupeReqFilterStream.ts (hunk 7 above)
Comment (danjm) [reply app/scripts/lib/createDupeReqFilterStream.ts]: d3264f5218

File: app/scripts/metamask-controller.js (hunk 8 above)
Comment (danjm) [reply app/scripts/metamask-controller.js]: d3264f5218

File: app/scripts/lib/createDupeReqFilterStream.ts (hunk 9 above)
Comment (danjm) [reply app/scripts/lib/createDupeReqFilterStream.ts]: d3264f5218

File: app/scripts/lib/createDupeReqFilterStream.test.ts (hunk 10 above)
Comment (danjm) [reply app/scripts/lib/createDupeReqFilterStream.test.ts]: d3264f5218

File: app/scripts/lib/createDupeReqFilterStream.test.ts (hunk 11 above)
Comment (danjm) [reply app/scripts/lib/createDupeReqFilterStream.test.ts]: d3264f5218

File: app/scripts/lib/createDupeReqFilterStream.test.ts [hunk 13]
```diff
@@ -0,0 +1,315 @@
...
+      .on('error', (err) => reject(err));
+
+    requests.forEach((request) => testStream.write(request));
+    testStream.end();
+
+    jest.advanceTimersByTime(advanceTimersTime);
+  });
+}
+
+describe('createDupeReqFilterStream', () => {
```
Comment (legobeat) [inline app/scripts/lib/createDupeReqFilterStream.test.ts:41]: (could) use a similar pattern to [`streams-utils` tests](https://github.com/MetaMask/metamask-extension/pull/24533/files#diff-2a5d1a92774c0fb88530dd20b5b5140f62b0ca423d5034c680e8d538a087df9fR48), if we want to ensure this works with streams instantiated from other implementations.

Comment (danjm) [reply app/scripts/lib/createDupeReqFilterStream.test.ts:41]: I don't _think_ we will ever need to use this with streams instantiated with other implementations, unless you, Mark or Erik can suggest a use case / need

Comment (legobeat) [reply app/scripts/lib/createDupeReqFilterStream.test.ts:41]: Do we _know_ that we don't already have versions mixed without full control of everything being single-implementation? (I'm not sure).

Also useful to ensure compatibility and prevent regressions when switching implementation (e.g. pending https://github.com/MetaMask/metamask-extension/pull/21934)

Comment (danjm) [reply app/scripts/lib/createDupeReqFilterStream.test.ts:41]: fair enough, I can update the tests to handle those cases.

File: app/scripts/lib/createDupeReqFilterStream.ts [hunk 14]
```diff
@@ -42,24 +43,26 @@ const makeExpirySet = () => {
...
       // JSON-RPC notifications have no ids; our only recourse is to let them through.
-      return next();
-    } else if (!seenRequestIds.add(req.id)) {
-      log.info(`RPC request with id "${req.id}" already seen.`);
-      return end();
-    }
-    return next();
-  };
+      const hasNoId = chunk.id === undefined;
+      const requestNotYetSeen = seenRequestIds.add(chunk.id);
```
[the remark this answers is not in the document]
Comment (danjm) [reply app/scripts/lib/createDupeReqFilterStream.ts:57]: It is okay to add undefined. Whether or not undefined is in the set, `hasNoId` will be `true` for all `undefined` ids, and so the below logic will have the same result (regardless of the presence of undefined in the set).

File: app/scripts/lib/createDupeReqFilterStream.test.ts (hunk 13 above)
Comment (danjm) [reply app/scripts/lib/createDupeReqFilterStream.test.ts:48]: Done in 44ac712dd2

**Review by rekmarks — APPROVED** [review]
LGTM!

**Review by legobeat — APPROVED** [review]

**Pull request merged**